src/App.tsx

```diff
@@ -1,93 +1,207 @@
-import React from 'react';
+import React, {
+  useCallback,
+  useContext, useEffect, useMemo, useRef, useState,
+} from 'react';
+
+import {
+  CSSTransition,
+  TransitionGroup,
+} from 'react-transition-group';
+
+import classNames from 'classnames';
+import {
+  createTodo, getTodos, deleteTodo, updateTodo,
+} from './api/todos';
+import { AuthContext } from './components/Auth/AuthContext';
+import { TodoList } from './components/TodoList';
+import { Footer } from './components/Footer';
+import { Todo } from './types/Todo';
+import { Error } from './components/Error';
+import { TodoCompleted } from './types/TodoCompleted';
+import { TodoTitle } from './types/TodoTitle';
+
+enum FilterType {
+  ALL = 'all',
+  ACTIVE = 'active',
+  COMPLETED = 'completed',
+}
 
 export const App: React.FC = () => {
+  const user = useContext(AuthContext);
+  const newTodoField = useRef<HTMLInputElement>(null);
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [query, setQuery] = useState('');
+  const [filter, setFilter] = useState<string>(FilterType.ALL);
+  const [errorType, setError] = useState('');
+  const [isLoading, setIsLoading] = useState(false);
+
+  const fireTimeOut = () => {
+    setTimeout(() => {
+      setError('');
+    }, 3000);
+  };
+
+  const addTodo = async () => {
+    if (!query.trim()) {
+      setError('Title can\'t be empty');
+      fireTimeOut();
+
+      return;
+    }
+
+    try {
+      const newTodo = await createTodo(user?.id, query);
+
+      setTodos((prevState) => {
+        return [...prevState, newTodo];
+      });
+    } catch (error) {
+      setError('Unable to add a todo');
+      fireTimeOut();
+    }
+  };
+
+  const removeTodo = useCallback(async (todoId: number) => {
+    try {
+      await deleteTodo(todoId);
+      setTodos(prev => prev.filter((x) => x.id !== todoId));
+    } catch (error) {
+      setError('Unable to delete a todo');
+      fireTimeOut();
+    }
+  }, []);
+
+  const changeTodo = useCallback(async (
+    todoId: number, object: TodoCompleted | TodoTitle,
+  ) => {
+    try {
+      const updetedTodo: Todo = await updateTodo(todoId, object);
+
+      setTodos(prev => (prev.map((x) => (x.id === todoId
+        ? updetedTodo
+        : x))
+      ));
+    } catch (error) {
+      setError('Unable to update a todo');
+      fireTimeOut();
+    }
+  }, []);
+
+  const fireChangeAllTodos = () => {
+    todos.forEach(todo => {
+      if (todos.every(everyTodo => everyTodo.completed === true)) {
+        changeTodo(todo.id, { completed: false });
+      } else if (todos.every(everyTodo => everyTodo.completed === false)) {
+        changeTodo(todo.id, { completed: true });
+      } else {
+        changeTodo(todo.id, { completed: true });
+      }
+    });
+  };
+
+  useEffect(() => {
+    // focus the element with `ref={newTodoField}`
+    getTodos(user?.id)
+      .then(res => setTodos(res))
+      .catch(() => {
+        setError('Unable to load todos');
+        fireTimeOut();
+      });
```

Comment:
move to a separate function and rewrite it on async/await. Just call it in useEffect()